contrib/src/main/scala/akka/stream/contrib/Valve.scala

```diff
@@ -0,0 +1,150 @@
+    */
+  def flip(mode: SwitchMode): Future[Boolean]
+}
+
+
+object Valve {
+  /**
+    * Factory for [[Valve]] instances.
+    *
+    * @tparam A input and output type
+    * @return [[Valve]] Opened instance
+    *
+    */
+  def apply[A](): Valve[A] = new Valve[A](SwitchMode.Open)
+
+  /**
+    * Factory for [[Valve]] instances.
+    *
+    * @tparam A input and output type
+    * @return [[Valve]] instance
+    *
+    */
+  def apply[A](mode: SwitchMode): Valve[A] = new Valve[A](mode)
+
+  /**
+    * Java API: Factory for [[Valve]] instances.
+    *
+    * @tparam A input and output type
+    * @return [[Valve]] Opened instance
+    */
+  def create[A](): Valve[A] = new Valve[A](SwitchMode.Open)
+
+  /**
+    * Java API: Factory for [[Valve]] instances.
+    *
+    * @tparam A input and output type
+    * @return [[Valve]] instance
+    */
+  def create[A](mode: SwitchMode): Valve[A] = new Valve[A](mode)
+
+}
+
+/**
+  * Materializes into a [[ValveSwitch]] which provides a the method flip that stops or restarts the flow of elements passing through the stage. As long as the valve is closed it will backpressure.
+  *
+  * Note that closing the valve could result in one element being buffered inside the stage, and if the stream completes or fails while being closed, that element may be lost.
+  *
+  * @param mode : state of the valve at the startup of the flow (by default Open)
+  */
+class Valve[A](mode: SwitchMode) extends GraphStageWithMaterializedValue[FlowShape[A, A], ValveSwitch] {
+
+  private val in: Inlet[A] = Inlet[A]("valve.in")
+  private val out: Outlet[A] = Outlet[A]("valve.out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ValveSwitch) = {
+    val logic = new ValveGraphStageLogic(shape, mode)
+    (logic, logic.switch)
+  }
+
+  private class ValveGraphStageLogic(shape: Shape, var mode: SwitchMode) extends GraphStageLogic(shape) with InHandler with OutHandler {
+    /**
+      * Buffer the element to push() if the valve was close if a push was requested
+      */
+    var bufferedElement = Option.empty[A]
+
+    val switch = new ValveSwitch {
+
+      override def flip(flipToMode: SwitchMode): Future[Boolean] = {
+        val promise = Promise[Boolean]()
+
+        val callback = getAsyncCallback[NotUsed] { _ =>
+          val succeed = mode match {
+            case _ if flipToMode == mode => false
+
+            case Open =>
+              mode = SwitchMode.Close
+              true
+              
+            case Close =>
+              bufferedElement.foreach(push(out, _))
+              bufferedElement = None
+
+              if (isAvailable(out)) {
+                // if a pull in has been requested we trigger pull(In)
+                pull(in)
+              }
+
+              mode = SwitchMode.Open
+              true
+          }
+
+          promise.success(succeed)
+        }
+
+        callback.invoke(NotUsed)
```

Comment:
You should have the Mode be the parameter of the async callback, then you would onyl have to allocate a single callback instead of one per flip-invocation.